README.md
# ValidatesHostname

- [Source](https://github.com/KimNorgaard/validates_hostname)
- [Bugs](https://github.com/KimNorgaard/validates_hostname/issues)

## Description

Extension to ActiveModel for validating hostnames and domain names.

## Features

- Adds validation for hostnames to ActiveModel
- Supports I18n for the error messages

## Installation

As a gem:

```bash
# in Gemfile
gem 'validates_hostname', '~> 2.0'

## Validations performed

- maximum length of hostname is 255 characters
- maximum length of each hostname label is 63 characters
- characters allowed in hostname labels are a-z, A-Z, 0-9 and hyphen
- labels do not begin or end with a hyphen
- labels do not consist of numeric values only

## Options

- option to allow for underscores in hostname labels
- option to require that the last label is a valid TLD (ie. require that the name is a FQDN)
- option to allow numeric values in the first label of the hostname (exception: the hostname cannot consist of a single numeric label)
- option to specify a list of valid TLDs
- options to allow for wildcard hostname in first label (for use with DNS)

See also http://www.zytrax.com/books/dns/apa/names.html

## How to use

Simple usage:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: true
end
```

With options:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { OPTIONS }
end
```

## Options and their defaults:

- `:allow_underscore => false`
- `:require_valid_tld => false`
- `:valid_tlds => Array` of allowed TLDs (can only be used with `:require_fqdn => true`)
- `:allow_numeric_hostname => false`

## Examples

Without options:

```ruby
class Record < ActiveRecord::Base
  validates :host, hostname: true
end

>> @record = Record.new(name: "horse")
>> @record.save
=> true

>> @record2 = Record.new(name: "_horse")
>> @record2.save
=> false
```

With `:allow_underscore`:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { allow_underscore: true }
end

>> @record3 = Record.new(name: "_horse")
>> @record3.save
=> true
```

With `:require_valid_tld`:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { require_valid_tld: true }
end

>> @record4 = Record.new(name: "horse")
>> @record4.save
=> false

>> @record5 = Record.new(name: "horse.com")
>> @record5.save
=> true
```

With `:valid_tlds`:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { require_valid_tld: true, valid_tlds: %w(com org net) }
end

>> @record6 = Record.new(name: "horse.info")
>> @record6.save
=> false
```

With `:allow_numeric_hostname`:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { allow_numeric_hostname: false }
end

>> @record7 = Record.new(name: "123.info")
>> @record7.save
=> false
```

With `:allow_wildcard_hostname`:

```ruby
class Record < ActiveRecord::Base
  validates :name, hostname: { allow_wildcard_hostname: true }
end

>> @record8 = Record.new(name: "*.123.info")
>> @record8.save
=> true
```

## Extra validators

A few extra validators are included.

### domainname

- sets `require_valid_tld => true`
- sets `allow_numeric_hostname => true`
- returns error if there is only one label and this label is numeric

### fqdn

- sets `require_valid_tld => true`

### wildcard

- sets `allow_wildcard_hostname => true`

## Error messages

Using the I18n system to define new defaults:

```yaml
en:
  errors:
    messages:
      invalid_label_length: "label must be between 1 and 63 characters long"
      label_begins_or_ends_with_hyphen: "label begins or ends with a hyphen"
      hostname_label_is_numeric: "unqualified hostname part cannot consist of numeric values only"
      single_numeric_hostname_label: "hostnames cannot consist of a single numeric label"
      label_contains_invalid_characters: "label contains invalid characters (valid characters: [%{valid_chars}])"
      invalid_hostname_length: "hostname must be between 1 and 255 characters long"
      tld_is_invalid: "tld of hostname is invalid"
```

The `%{valid_chars}` signifies the range of valid characters allowed in labels.

It is highly recommended you use the I18n system for error messages.

## Maintainers

- [Kim Nørgaard](mailto:jasen@jasen.dk)

## License

Copyright (c) 2009-2025 Kim Norgaard, released under the MIT license.


CHANGELOG
2009-06-18 - Kim Nørgaard <jasen@jasen.dk>
  * Initial commit


lib/validates_as_hostname.rb
module PAK
  module ValidatesAsHostname

  # List from IANA: http://www.iana.org/domains/root/db/
  ALLOWED_TLDS = %w(
    ac ad ae aero af ag ai al am an ao aq ar arpa as asia at au aw ax az 
    ba bb bd be bf bg bh bi biz bj bl bm bn bo br bs bt bv bw by bz 
    ca cat cc cd cf cg ch ci ck cl cm cn co com coop cr cu cv cx cy cz 
    de dj dk dm do dz 
    ec edu ee eg eh er es et eu 
    fi fj fk fm fo fr 
    ga gb gd ge gf gg gh gi gl gm gn gov gp gq gr gs gt gu gw gy 
    hk hm hn hr ht hu 
    id ie il im in info int io iq ir is it 
    je jm jo jobs jp 
    ke kg kh ki km kn kp kr kw ky kz 
    la lb lc li lk lr ls lt lu lv ly 
    ma mc md me mf mg mh mil mk ml mm mn mo mobi mp mq mr ms mt mu museum 
    mv mw mx my mz 
    na name nc ne net nf ng ni nl no np nr nu nz 
    om org 
    pa pe pf pg ph pk pl pm pn pr pro ps pt pw py 
    qa 
    re ro rs ru rw 
    sa sb sc sd se sg sh si sj sk sl sm sn so sr st su sv sy sz 
    tc td tel tf tg th tj tk tl tm tn to tp tr travel tt tv tw tz 
    ua ug uk um us uy uz 
    va vc ve vg vi vn vu 
    wf ws 
    ye yt yu 
    za zm zw
  )


  def validates_as_domainname(*attr_names)
    options = {
      :require_valid_tld        => true,
      :allow_numeric_hostname   => true
    }.merge(attr_names.last.is_a?(Hash) ? attr_names.pop : {})

    validates_as_hostname(*attr_names + [options])
    
    # each +value+ is a full hostname
    # split these into labels and do various checks
    validates_each(attr_names, options) do |r, attr_name, value|
      if value.is_a?(String)
        labels = value.split '.'
        labels.each do |label|
          # CHECK 1: if there is only one label it cannot be numeric even
          #          though numeric hostnames are allowed
          if options[:allow_numeric_hostname] == true
            is_numeric_only = (
              (
                Integer(labels[0]) rescue false
              ) ? true : false
            )
            if is_numeric_only and labels.size == 1
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.single_numeric_hostname_label',
                  :default => 'hostnames cannot consist of a single ' \
                            + 'numeric label'
                )
              )
            end
          end
        end
      end
    end
  end


  def validates_as_fqdn(*attr_names)
    options = {
      :require_valid_tld        => true,
    }.merge(attr_names.last.is_a?(Hash) ? attr_names.pop : {})

    validates_as_hostname(*attr_names + [options])
  end


  def validates_as_hostname(*attr_names)
      options = {
        :allow_underscore        => false,
        :require_valid_tld       => false,
        :valid_tlds              => ALLOWED_TLDS,
        :allow_numeric_hostname  => false
      }.merge(attr_names.last.is_a?(Hash) ? attr_names.pop : {})


      # maximum hostname length: 255 characters
      validates_length_of *attr_names + [{
        :in => 1..255,
        :message => I18n.t(
                      'validates_as_hostname.invalid_hostname_length',
                      :default => 'hostname must be between 1 and 255 ' \
                                + 'characters long'
        )
      }.merge(options)]

      # each +value+ is a full hostname
      # split these into labels and do various checks
      validates_each(attr_names, options) do |r, attr_name, value|
        if value.is_a?(String)
          labels = value.split '.'
          labels.each do |label|
            # CHECK 1: hostname label cannot be longer than 63 characters
            if label.length > 63
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.invalid_label_length',
                  :default => 'label must be between 1 and 63 characters ' \
                            + 'long'
                )
              )
            end

            # CHECK 2: hostname label cannot begin or end with hyphen
            if label =~ /^[-]/i or label =~ /[-]$/
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.label_begins_or_ends_with_hyphen',
                  :default => 'label begins or ends with a hyphen'
                )
              )
            end

            # CHECK 3: hostname can only contain characters: 
            #          a-z, 0-9, hyphen, optionally underscore
            valid_chars = 'a-z0-9\-'
            valid_chars << '_' if options[:allow_underscore] == true
            unless label =~ /^[#{valid_chars}]+$/i
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.label_contains_invalid_characters',
                  :default => "label contains invalid characters (valid " \
                            + "characters: [#{valid_chars}])"
                )
              ) 
            end
          end

          # CHECK 4: the unqualified hostname portion cannot consist of 
          #          numeric values only
          if options[:allow_numeric_hostname] == false
            is_numeric_only = (
              (
                Integer(labels[0]) rescue false
              ) ? true : false
            )
            if is_numeric_only
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.hostname_label_is_numeric',
                  :default => 'unqualified hostname part cannot consist ' \
                            + 'of numeric values only'
                )
              )
            end
          end

          # CHECK 5: in order to be fully qualified, the full hostnames
          #          TLD must be valid
          if options[:require_valid_tld] == true
            has_tld = options[:valid_tlds].select {
              |tld| tld =~ /^#{labels.last}$/i
            }.empty? ? false : true
            unless has_tld
              r.errors.add( attr_name,
                I18n.t(
                  'validates_as_hostname.hostname_is_not_fqdn',
                  :default => 'hostname is not a fully qualified domain name'
                )
              )
            end
          end
        end
      end
  end
  end
end

ActiveRecord::Base.extend PAK::ValidatesAsHostname
